### PR TITLE
Add logging to Network

### PR DESCRIPTION
**Pull request by kidroca**
@tgolen @marcaaron 

### Details

Added Logging to `Network.js` 
Added retry limit to failed `shouldRetry` requests 

### Fixed Issues

$ https://github.com/Expensify/App/issues/7666

### Tests

Find a way to make requests fail like switch network off in the middle of the request (it's hard to nail this)
Observe captured logs 
A request should not be retried more than 10 times

An easier (but lengthily) way to simulate this is 
1. Turn airplane mode on (or be offline)
2. Write a message 
3. Quit the app (terminate it) (you can resume connectivity after)
4. Wait for 2 hours so you're auth token expires
5. Launch the app
6. Observe logs for persisted request being retried (server logs)

#### 2 AddComment request being made and retried 
```
[info] Making API request - {"command":"Report_AddComment","type":"post","shouldUseSecure":false}
[info] Making API request - {"command":"Report_AddComment","type":"post","shouldUseSecure":false}
[info] Persisted request failed - {"retryCount":1,"command":"Report_AddComment","error":"Persisted request failed due to jsonCode: 407"}
[info] Persisted request failed - {"retryCount":1,"command":"Report_AddComment","error":"Persisted request failed due to jsonCode: 407"}
[info] Making API request - {"command":"Report_AddComment","type":"post","shouldUseSecure":false}
[info] Making API request - {"command":"Report_AddComment","type":"post","shouldUseSecure":false}
[info] Persisted request failed - {"retryCount":2,"command":"Report_AddComment","error":"Persisted request failed due to jsonCode: 407"}
[info] Persisted request failed - {"retryCount":2,"command":"Report_AddComment","error":"Persisted request failed due to jsonCode: 407"}
[info] Making API request - {"command":"Report_AddComment","type":"post","shouldUseSecure":false}
[info] Making API request - {"command":"Report_AddComment","type":"post","shouldUseSecure":false}
```

### QA Steps

Not sure whether QA can help here, maybe just check for regression related to networking
E.g. sending a message offline should still send the message after you resume connectivity 
It should be sent only one time

### Tested On

- [x] Web
- [x] Mobile Web
- ❌Desktop - untestable due to CORS issue
- [x] iOS
- [x] Android

#### Web

image

#### Mobile Web

image

#### iOS

image
image

#### Android

image

### CODE REVIEW

**Comment by kidroca** [comment]
I was able to test this by manually invalidating my session.authToken by merging this to `session` 
```
{authToken: 'simulate expired token'}
```

**Comment by kidroca** [comment]
I'm yet to test on the rest of the platforms but want to let you review whether this direction is the one we want to go
Will test and post screenshots in an hour or so

File: src/libs/Network.js [hunk 1]
```diff
@@ -56,13 +57,16 @@ function processPersistedRequestsQueue() {
...
             if (response.jsonCode !== CONST.HTTP_STATUS_CODE.SUCCESS) {
-                throw new Error('Persisted request failed');
+                throw new Error(`Persisted request failed due to jsonCode: ${response.jsonCode}`);
             }
 
             NetworkRequestQueue.removeRetryableRequest(request);
         })
-        .catch(() => {
+        .catch((error) => {
+            const Log = getLogger();
```
Comment (tgolen) [inline src/libs/Network.js:66]: Maybe just move this to the very top of the file so that it's only called once?

**Review by tgolen — CHANGES_REQUESTED** [review]
This looks like a good place to start with the logging, yeah.

File: src/libs/Network.js (hunk 1 above)
Comment (kidroca) [reply src/libs/Network.js:66]: It wouldn't work 
image

Another script (API) first reads (imports) the `Network.js` script and evaluates all to level expression e.g. `const Log = getLogger()` 
And only then assigns value to be returned by `getLogger()`

Comment (tgolen) [reply src/libs/Network.js:66]: Ah, OK. Makes sense

**Comment by kidroca** [comment]
This is what happens on Desktop ATM:

1. A request fails due to CORS
2. It goes to the `catch` block
3. We retry a bunch of times
4. Retires themselves are logging information and this creates a new Log request that fails again due to CORS\
5. This creates a loop

https://user-images.githubusercontent.com/12156624/153456883-b9e3d4cf-9a6a-48a9-ad70-c0aa34b7bfd9.mov

```
[info] A retrieable request failed - {"retryCount":9,"command":"Log","error":"Failed to fetch"}
localhost/:1 Access to fetch at 'https://www.expensify.com/api?command=Log' from origin 'http://localhost:8080' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
HttpUtils.js:23 
 POST https://www.expensify.com/api?command=Log net::ERR_FAILED 200
Log.js:54 [info] A retrieable request failed - {"retryCount":10,"command":"Log","error":"Failed to fetch"}
Log.js:54 [alrt] Request was retried too many times with no success. No more retries left - {"stack":"\"Error\\n    at Logger.alert (http://localhost:8080/app-20468ac….bundle.js:81111:43)\\n    at http://localhost:8080/app-20468ac….bundle.js:346466:21\""}
API.js:196 [API] There was an error in the Log API command, unable to log to server! TypeError: Failed to fetch
    at processHTTPRequest (HttpUtils.js:23:12)
    at Object.xhr (HttpUtils.js:47:12)
    at processRequest (Network.js:46:22)
    at Network.js:231:9
    at Function.each (each.js:14:1)
    at processNetworkRequestQueue (Network.js:221:7)
```

I'm not sure if we should do something about this, because if you are getting CORS errors the app is in a useless state anyway

Note: as you can see there's no way to distinct (in code) that this is a CORS error it's simply `TypeError: Failed to fetch`. The same thing you would get when the request fails mid-flight

File: src/libs/Network.js (hunk 1 above)
Comment (kidroca) [reply src/libs/Network.js]: Removed `const Log = getLogger()`
Replaced usages to be like `getLogger().info(...)` 

Is that better?

**Review by alex-mechler — COMMENTED** [review]
This looks good to me, but you have conflicts

**Comment by kidroca** [comment]
Merge conflict fixed 

Still CORS errors on Desktop though... https://expensify.slack.com/archives/C01GTK53T8Q/p1644532587614239?thread_ts=1644514188.692809&cid=C01GTK53T8Q

**Review by tgolen — APPROVED** [review]

**Pull request merged**